ztest: Update userspace tests in new API

yperess opened the pull request:
Update the new API to use K_USER as the flags for both
CONFIG_USERSPACE and CONFIG_TEST_USERSPACE.

Fixes #44108

Signed-off-by: Yuval Peress <peress@google.com>

Commit:
ztest: Fix userspace ztests in new API

Update the new API to use K_USER as the flags for both
CONFIG_USERSPACE and CONFIG_TEST_USERSPACE. Also, fix the linker
script to properly include the suites, tests, and rules.

Fixes #44108

Signed-off-by: Yuval Peress <peress@google.com>

henrikbrixandersen reviewed (dismissed):
Thanks. Please add a test case for the new API running in userspace.

yperess commented:
@henrikbrixandersen I ran into a bit of a snag with the test. When adding `CONFIG_TEST_USERSPACE` I can't get the test `zassert_true(k_is_user_context(), NULL);` to pass because of the logic in `k_is_user_context()` which doesn't factor in that config:

```
static inline bool k_is_user_context(void)
{
#ifdef CONFIG_USERSPACE
	return arch_is_user_context();
#else
	return false;
#endif
}
```

I've included the tests for `CONFIG_USERSPACE`, but I'm wondering if there's more work to get `CONFIG_TEST_USERSPACE` to work as you've intended?

yperess commented:
@henrikbrixandersen I'm so far unable to figure out why some of these tests pass. I ran `tests/ztest/error_hook` tests and verified that `k_is_user_context()` is returning true there so I'm sure that the function works, I also checked that the `thread_options` in the test struct is using `K_USER`. Any thoughts?

henrikbrixandersen commented:
> @henrikbrixandersen I'm so far unable to figure out why some of these tests pass. I ran `tests/ztest/error_hook` tests and verified that `k_is_user_context()` is returning true there so I'm sure that the function works, I also checked that the `thread_options` in the test struct is using `K_USER`. Any thoughts?

No, sorry. I was unable to get the new ztest API to work in userspace myself, which is why I asked for tests to be added. I ended up reverting to the "classic" ztest API due to this.

yperess commented:
Looking at this some more the error:
```
E: Page fault at address 0x40005054 (error code 0x4)
E: Linear address not present in page tables
E: Access violation: user thread not allowed to read
E: PTE: not present
E: EAX: 0x4002249d, EBX: 0x40005050, ECX: 0x40005050, EDX: 0x4000502c
E: ESI: 0x00000000, EDI: 0x40039fe8, EBP: 0x40039fcc, ESP: 0x40039fc4
E: EFLAGS: 0x00000202 CS: 0x002b CR3: 0x00133720
E: call trace:
E: EIP: 0x400224b2
E:      0x40020005 (0x4000502c)
E: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0
E: Current thread: 0x40032120 (test_catch_assert_fail)
Caught system error -- reason 0 0
Fatal error was unexpected, aborting...
```
seems to be related to https://github.com/zephyrproject-rtos/zephyr/issues/21317 so the answer possibly exists in the commits the resolved that issue. Right now the thread for the test is getting initialized as expected, but we're never actually getting to the thread's callback.

yperess commented:
@andrewboie looking at https://github.com/zephyrproject-rtos/zephyr/pull/44562#issuecomment-1093277128 and that it's similar to the issue you were running into a while back, do you know (or have ideas for) how to make sure that the `ztest_unit_test` objects are available to the userspace threads?

yperess commented:
After some more digging I'm certain that the issue is in how https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/linker/common-ram.ld#L143 are done, I've update this PR to something that seems closer to the right solution but I'm getting a syntax error that I haven't been able to figure out yet.

yperess commented:
> @dcpleung can you please take a look?

https://github.com/zephyrproject-rtos/zephyr/pull/44562/commits/f5adfe36320da7af348a5ccbfbe8566344626947 is the relevant commit to what I was trying to do (though it's basically a shot in the dark)

dcpleung commented on subsys/testsuite/include/ztest.ld:
```diff
@@ -1,28 +0,0 @@
-
-SECTIONS
-{
-	.data.ztest_suite_node_area : ALIGN(4)
```
I think you still need to copy this line over.

yperess replied:
I tried that just now but when I run `readelf` on the binary I was seeing that the sections were placed in the wrong location. I've update the PR just now with something that builds at least and is cleaner (again the 3rd commit is the one I'm trying to debug).

yperess replied:
@dcpleung `if (suite->before) {` appears to be the offending instruction but I'm not seeing why it would do so. I added more details to the discord chat thread

yperess replied:
I got a bit further by using `RAM_SECTIONS` instead of the data sections

yperess replied:
Updated again, more boards now pass, but I'm still seeing the same error in other boards. It seems to believe that the data in the test suite is not acceissble.

Commit:
ztest: add userspace tests to base tests

Add a test that verifies that ZTEST_USER is actually being run in
user context.

Signed-off-by: Yuval Peress <peress@google.com>

Commit:
ztest: migrate error_hook tests to new API

Update the tests to use the new ztest API.

Signed-off-by: Yuval Peress <peress@google.com>

yperess commented:
@henrikbrixandersen @dcpleung I got things 99% working, the only thing that fails now is riscv64 architectures and I believe it's caused by an alignment issue that I'm not quite seeing. I'll continue to hammer on this tomorrow.

Commit:
soc: xtensa: fix missing rodata section in intel_adsp

The common linker script for cavs_v?? was missing an include for the
snippets-rodata.ld file which is needed for any applications using the
`zephyr_linker_source(RODATA <linker_script>)` cmake function.

Signed-off-by: Yuval Peress <peress@google.com>

Commit:
soc: xtensa: fix missing rodata section in intel_s1000

The linker script for the intel_s1000 was missing an include for the
snippets-rodata.ld file which is needed for any applications using the
`zephyr_linker_sources(RODATA <linker_script>)` cmake function.

Signed-off-by: Yuval Peress <peress@google.com>

yperess commented:
@henrikbrixandersen @dcpleung this is now working and verified

keith-zephyr commented on soc/xtensa/intel_adsp/common/include/cavs-link.ld:
```diff
@@ -274,6 +274,11 @@ SECTIONS {
+
+    . = ALIGN(4);
+    #include <snippets-rodata.ld>
+    . = ALIGN(4);
```
Are both alignments required?  If they are actually required, I prefer you remove them (especially the 2nd one since there was no alignment previously for the __XT_EXCEPTION_TABLE__

andyross replied:
Arguably even if they are required, they belong in the snippet definitions anyway.  Zephyr linkage management gets a little wonky in places.  But yeah, it's hard to imagine that a 1-word alignment is being defeated somehow.

The core change here is for sure needed though; the cAVS linker script went through a big rework recently and clearly I missed this bit.

yperess replied:
Done

keith-zephyr commented on soc/xtensa/intel_s1000/linker.ld:
```diff
@@ -368,6 +368,11 @@ SECTIONS
     *(.gnu.linkonce.r.*)
     *(.rodata1)
+
+    . = ALIGN(4);
```
Same thing here.  If the alignment isn't required, please remove.

yperess replied:
Done

keith-zephyr commented on subsys/testsuite/include/ztest_unittest.ld:
```diff
@@ -0,0 +1,28 @@
+
+SECTIONS
+{
+	.data.ztest_suite_node_area : ALIGN(4)
```
Does this need to use _ZTEST_ALIGN?

yperess replied:
The `ztest_unittest.ld` is used in a completely different build which only runs native posix. So there's no need for it to be anything other than 4.

henrikbrixandersen commented:
> @henrikbrixandersen @dcpleung this is now working and verified

Great, thank you. I'd love to see additional tests added for `ZTEST_USER_F()` at some point, but it's not a blocker for me. I don't know enough of the Zephyr linker scripts to feel comfortable approving this PR, but it LGTM.

yperess commented:
Removed additional `ALIGN(4)` calls pointer out by Keith.

yperess commented:
@keith-zephyr @dcpleung Can I get another pass at review?

yperess commented:
@henrikbrixandersen It looks like you need to approve this as well because you requested changes.
